### PR TITLE
klientctl: cobra CLI infrastructure implementation

### PR DESCRIPTION
**Pull request by ppknap**
This PR introduces `spf13/cobra` CLI to `kd`

It defines all middlewares that are required for correct KD behavior.

⚠️ This PR doesn't change current `kd` CLI so it can be merged and used for testing. In order to use `cobra` cli `cobra` tag is required during `klientctl` installation/compilation.

Currently, there is only one command implemented(as a POC) `kd version`.

## Architecture

Cobra commands usually are stored in `commands` directory. Which contains command constructors and implementation.

Inside `commands` directory, we store `cli` package(similar to `docker` but with proper package tree). The `cli` package defines command *middlewares* - modifiers to the `RunE` functions that adds common functionality. Despite middlewares, `cli` package also contains `CLI` *type*, which object is passed to all commands and stores data streams, loggers, metrics client and all other data structures that might be used by command implementations. `CLI` type is like `context.Context` but it stores persistent(per process) objects.

`main.go` is kept as small as possible. All other functionality is added per commands via *middlewares*.

Moreover:
* there are no global objects/flags used by CLI.
* any input/output goes through `*cli.CLI` object. No need to run separate process during tests.
* any mocking should be done by modifying `*cli.CLI` object. 
* aliased commands should have its `New*CommandAlias` constructor to work properly with `metric` middleware.

The architecture is based on `cobra`'s best practices and takes some ideas from `github.com/rkt/rkt`, `github.com/moby/moby`(docker), and `github.com/spf13/hugo` packages.

## Motivation and Context
Better `kd` UX and pre-requirement for command autocompletion.

## How Has This Been Tested?
Manually.

## Screenshots (if appropriate):
https://asciinema.org/a/f23l0giazbucqf2676vrlfr3z

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: go/src/koding/klientctl/commands/cli/cli.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	if w == nil {
+		w = ioutil.Discard
+	}
+
+	handler := logging.NewWriterHandler(w)
+
+	// Make handler writer accept all incoming log messages.
+	handler.SetLevel(logging.DEBUG)
+
+	logger := logging.NewLogger("kd")
```
Comment (cihangir) [inline go/src/koding/klientctl/commands/cli/cli.go]: what about NewCustom?

File: go/src/koding/klientctl/commands/cli/cli.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+func (c *CLI) Log() logging.Logger {
+	if c.log != nil {
+		return c.log
+	}
+
+	return newLogger(nil)
+}
+
+// Metrics returns metrics client or nil if not enabled.
+func (c *CLI) Metrics() *metrics.Metrics {
```
Comment (cihangir) [inline go/src/koding/klientctl/commands/cli/cli.go]: Seems like we can export c.m as c.Metrics

File: go/src/koding/klientctl/commands/version/cmd.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		}
+
+		v.Latest, _ = config.LatestKDVersionNum()
+		if opts.jsonOutput {
+			cli.PrintJSON(c.Out(), v)
+			return nil
+		}
+
+		fmt.Fprintf(c.Out(), "Installed Version: %s\n", getReadableVersion(v.Installed))
+		fmt.Fprintf(c.Out(), "Latest Version: %s\n", getReadableVersion(v.Latest))
```
Comment (cihangir) [inline go/src/koding/klientctl/commands/version/cmd.go]: should we omit this line if Latest is empty?

**Review by cihangir — APPROVED** [review]
LGTM 👍

**Review by rjeczalik — APPROVED** [review]

File: go/src/koding/klientctl/commands/cli/cli.go (hunk 2 above)
Comment (ppknap) [inline go/src/koding/klientctl/commands/cli/cli.go]: Yes, we can. However, I'm still not sure if this becomes an interface (like in Docker).

Moreover, I wanted to not add a public member to the structure which has only getters(consistency).

**Pull request merged**